[JSC] Refine B3 and masm tests for EXTR pattern detection and application

hyjorc1 opened the pull request:
#### 084df28f4fc4fcadd396cf7f39f16374b1a524a8
<pre>
[JSC] Refine B3 and masm tests for EXTR pattern detection and application
<a href="https://bugs.webkit.org/show_bug.cgi?id=264984">https://bugs.webkit.org/show_bug.cgi?id=264984</a>
<a href="https://rdar.apple.com/118532295">rdar://118532295</a>

Reviewed by Keith Miller.

Previously we introduced the pattern detection and application
of EXTR when B3 is lowering to AIR. This patch does two things:
1. Refine the corresponding B3 and masm tests for EXTR.
2. Add a sanity check and more comments for EXTR pattern detection and application in B3.

* Source/JavaScriptCore/assembler/testmasm.cpp:
(JSC::testExtractRegister32):
(JSC::testExtractRegister64):
* Source/JavaScriptCore/b3/B3LowerToAir.cpp:
* Source/JavaScriptCore/b3/testb3_2.cpp:
(testExtractRegister32):
(testExtractRegister64):

Canonical link: <a href="https://commits.webkit.org/274149@main">https://commits.webkit.org/274149@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2bec5500857dae9f95ecf613e392708cfa5366b9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/37910 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/16818 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/40172 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/40453 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/33731 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/19513 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/14118 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/32058 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/38482 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/14170 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/33209 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12365 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/12325 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/33890 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/41725 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/31609 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/34396 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/34372 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38191 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/37608 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/12908 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/10551 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36371 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/14484 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/44534 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/13341 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/9118 "Found 1 new JSC stress test failure: wasm.yaml/wasm/gc/bug266127.js.wasm-eager (failure)") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4942 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/13936 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hyjorc1 commented on Source/JavaScriptCore/b3/B3LowerToAir.cpp:
```diff
@@ -3379,9 +3388,12 @@ class LowerToAir {
                 uint64_t highWidth = highWidthValue->asInt();
                 uint64_t lowWidth = lowWidthValue->asInt();
                 uint64_t datasize = opcode == ExtractRegister32 ? 32 : 64;
+                // Note that when `lowWidth == highWidth` we cannot turn it to `MOV Rd Rn` since
```
Oops, it should be `lowWidth == datasize`.

kmiller68 commented on Source/JavaScriptCore/assembler/testmasm.cpp:
```diff
@@ -1784,12 +1784,11 @@ void testExtractSignedBitfield64()
         for (auto m : int32Operands()) {
-            for (auto lsb : imms) {
+            for (uint32_t lsb = 0; lsb < datasize; ++lsb) {
                 if (lsb < datasize) {
```
I don't think you need this anymore.

kmiller68 commented on Source/JavaScriptCore/b3/testb3_2.cpp:
```diff
@@ -4899,100 +4897,96 @@ void testExtractRegister32()
 
         auto code = compileProc(proc);
-        if (isARM64() && lowWidth > 0)
+        if (checkEmittedEXTR && isARM64() && 0 < lowWidth && lowWidth < 32)
```
Why do we have `checkEmittedEXTR`? Shouldn't we always be checking for the instruction when we think we should have it?

hyjorc1 replied:
Certain `lowWidth` values can trigger other optimizations in `B3ReduceStrength.cpp`, which can break this `EXTR` pattern. Then, `checkEmittedEXTR` is introduced since not every test case can emit `EXTR`.

kmiller68 commented on Source/JavaScriptCore/b3/testb3_2.cpp:
```diff
@@ -4899,100 +4897,96 @@ void testExtractRegister32()
 
         auto code = compileProc(proc);
-        if (isARM64() && lowWidth > 0)
+        if (checkEmittedEXTR && isARM64() && 0 < lowWidth && lowWidth < 64)
```
Ditto.

hyjorc1 replied:
Certain `lowWidth` values can trigger other optimizations in `B3ReduceStrength.cpp`, which can break this `EXTR` pattern. Then, `checkEmittedEXTR` is introduced since not every test case can emit `EXTR`.

kmiller68 reviewed (approved):
r=me with nits.

hyjorc1 committed:
[JSC] Refine B3 and masm tests for EXTR pattern detection and application
https://bugs.webkit.org/show_bug.cgi?id=264984
rdar://118532295

Reviewed by Keith Miller.

Previously we introduced the pattern detection and application
of EXTR when B3 is lowering to AIR. This patch does two things:
1. Refine the corresponding B3 and masm tests for EXTR.
2. Add a sanity check and more comments for EXTR pattern detection and application in B3.

* Source/JavaScriptCore/assembler/testmasm.cpp:
(JSC::testExtractRegister32):
(JSC::testExtractRegister64):
* Source/JavaScriptCore/b3/B3LowerToAir.cpp:
* Source/JavaScriptCore/b3/testb3_2.cpp:
(testExtractRegister32):
(testExtractRegister64):

Canonical link: https://commits.webkit.org/274149@main
